### PR TITLE
Bugfix/libwebsocket dtor deadlock

### PR DESCRIPTION
**Pull request by AssemblyJohn**
## Describe your changes

- Better split code and thread responsibilities
- Added a generic safe queue to simplify count of conditional variables and mutexes used
- Cleared the connection_mutex usage in websocket client threads
- Modified reconnect logic to internal reconnect logic on the websocket thread for better stability
- The `connect` in websocket is not sync any more, but only kickstarts the connections attempts

## Issue ticket number and link

Core CI: https://github.com/EVerest/everest-core/pull/984

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

**Comment by AssemblyJohn** [comment]
The integration tests have been run, the unit tests have been run and received feedback for required behavior for some corner-cases. Everything passes, so I am requesting a review with other possible problems that we might encounter with this code reorganization: @hikinggrass @marcemmers @Pietfried @maaikez

**Review by Pietfried — APPROVED** [review]

File: lib/ocpp/common/websocket/websocket_libwebsockets.cpp [hunk 1]
```diff
@@ -604,429 +678,411 @@ void WebsocketLibwebsockets::client_loop() {
...
 
-// Will be called from external threads as well
-bool WebsocketLibwebsockets::connect() {
-    if (!this->initialized()) {
-        return false;
-    }
+void WebsocketLibwebsockets::clear_all_queues() {
+    message_queue.clear();
+    recv_buffered_message.clear();
+    recv_message_queue.clear();
```
Comment (marcemmers) [inline lib/ocpp/common/websocket/websocket_libwebsockets.cpp:877]:
```suggestion
    this->message_queue.clear();
    this->recv_buffered_message.clear();
    this->recv_message_queue.clear();
```

Looks like we don't do this throughout the whole file anyway.

**Review by marcemmers — CHANGES_REQUESTED** [review]
Ran out of time, these are my findings until now. Mostly small ones

File: include/ocpp/common/safe_queue.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        std::lock_guard lock(mutex);
+        return queue.empty();
+    }
+
+    inline safe_queue_reference front() const {
+        std::lock_guard lock(mutex);
+        return queue.front();
+    }
+
+    inline safe_queue_const_reference front() {
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:29]: I will try to see if I can const it, should make the mutex 'mutable'.

File: include/ocpp/common/safe_queue.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        if (seconds > 0) {
+            cv.wait_for(lock, std::chrono::seconds(seconds), [&]() { return pred(); });
+        } else {
+            cv.wait(lock, [&]() { return pred(); });
+        }
+    }
+
+    /// \brief Notifies a single waiting thread to wake up
+    inline void notify_waiting_thread() {
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:107]: Any suggestions? Expose the `cv`?

File: include/ocpp/common/safe_queue.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    inline void clear() {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        std::queue<T> empty;
+        empty.swap(queue);
+    }
+
+    /// \brief Waits seconds for the queue to receive an element
+    /// \param seconds Count of seconds to wait, pass in a value <= 0 to wait indefinitely
+    inline void wait_on_queue(int seconds = -1) {
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:74]: Makes sense, especially since there are waits involved, milli precision is better.

File: include/ocpp/common/safe_queue.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcemmers) [reply include/ocpp/common/safe_queue.hpp:107]: I am not sure, it's probably not an easy thing to do right so it's probably best to continue with this for now.

To elaborate on my concern. There are 2 things we are doing; 
1. We have a queue that we need to protect for multi threaded access, that is done in this class. 
2. We have a thread running that is waiting for input from the queue or input from other places. This is now done half in this class (the `cv`) and half in the thread part of the websocket. So either we fully move the thread into this class and handle it only in here, or we move the `cv` out of this one. This has more implications I know, that's why I think its best to leave it as is for now.

[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:107]: I do understand the concerns. However, I've thought at something. The problem is that not always the queue knows when to wake up the waiting thread. When we push a message? When we pop a message? Both? That's why the 'notify_waiting_thread' could be removed and replaced with a C++ template policy enum that specifies the used case. 

For example:

`
enum ESafeQueueWakeupPolicy {
WAKE_UP_NEVER,
WAKE_UP_ON_POP,
WAKE_UP_ON_PUSH,
WAKE_UP_ALWAYS
};
`

Of course that adds the complexity of having to know what the queue will do.

For moving the thread in the queue, I've also thought at that but I wouldn't do it in the `SafeQueue` class, I'd create a `QueueWorkerThread` that contains both the `std::thread` and an internal `SafeQueue` and that receives as a ctor a lambda worked parameter or something along those lines.

However at the moment it seems the queue wakeup policy makes the most sense.

[the remark this answers is not in the document]
Comment (marcemmers) [reply include/ocpp/common/safe_queue.hpp:107]: I am not really sure how that would look yet, might be a good option. At the same time, I understood this needs to be merged soon for some kind of release so might be wise to just accept that this is some technical debt and accept it for now?

[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:107]: It would not be that complicated (~30 mins of work), and it would remove the 'notify_waiting_thread()' function better separating functionality as you've said. That will give us a head-start when we will properly fix the other technical debt, the proper management of the thread that is attached to this queue.

File: lib/ocpp/common/websocket/websocket_libwebsockets.cpp [hunk 5]
```diff
@@ -81,80 +87,179 @@ struct ConnectionData {
         owner = nullptr;
     }
 
-    void bind_thread(std::thread::id id) {
-        lws_thread_id = id;
+public:
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply lib/ocpp/common/websocket/websocket_libwebsockets.cpp:90]: I use them as separators, especially when the class tends to become a bit large, but this doesn't seem to go with the project's style.

File: include/ocpp/common/safe_queue.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:138]: I've implemented here a queue policy on when to wake up. However I am not sure that it is the best idea. Maybe we should notify the thread always to wake up on an operation?

https://github.com/EVerest/libocpp/pull/896/commits/4ac4485db2d6091c80e28a9b0280d74570a0af14

**Review by marcemmers — APPROVED** [review]

File: include/ocpp/common/safe_queue.hpp [hunk 6]
```diff
@@ -9,8 +9,19 @@
...
 
+enum class EThreadNotifyPolicy {
+    // Never notify the waiting thread
+    ThreadNotify_Never,
+    // Notify the waiting thread when we push an element in the queue
+    ThreadNotify_Push,
+    // Notify the waiting thread when we pop an element from the queue
+    ThreadNotify_Pop,
+    // Always notify a waiting thread on all operations
+    ThreadNotify_Always,
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:20]: I was referring to the idea behind it. Not sure it makes sense? Maybe we should always notify threads on all operations? @marcemmers @Pietfried what are your thoughts?

File: include/ocpp/common/safe_queue.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pietfried) [reply include/ocpp/common/safe_queue.hpp:138]: I dont see any drawback with notifying on each queue operation and then we could omit the policy. I would suggest to accept the changes for now.

Of course one could reduce the scope and complexity of the `SafeQueue` class to the basic queue operations, move the cv out of it and pass e.g. a lambda triggered on every operation (or even based on a passed policy), and we can discuss this afterwards.

But in the current design of the libwebsockets impl the logic that uses the queues requires the notifications and this leads to the fact that the SafeQueue always need the notify mechanism one or the other way.  Therefore I'm also fine with the current implementation since the `SafeQueue` is currently only used here

File: include/ocpp/common/safe_queue.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Pietfried) [reply include/ocpp/common/safe_queue.hpp:20]: This was just a stylistic hint, I didnt verify the idea when I added the previous comment ;)

**Review by Pietfried — APPROVED** [review]

File: include/ocpp/common/safe_queue.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply include/ocpp/common/safe_queue.hpp:116]: At the moment I've removed the policy and will just notify the thread on all operations. The old behavior was that the CV is out of the class, however that added a lot of failure points in the libwebsocket class that always required to call that CV notify, since in there we have 3 queues.

Maybe the lambda triggered on each operation could be the best idea, but I'd say to postpone it until we need the SafeQueue in another class that is not the websocket.
